crates/bevy_render/src/allocator.rs

```diff
@@ -0,0 +1,560 @@
+//! An allocator that divides up vertex and index buffers so that multiple
+//! meshes can be packed together.
+//!
+//! The underlying allocation algorithm is [`offset-allocator`], which is a port
+//! of [Sebastian Aaltonen's `OffsetAllocator`]. It's a fast, simple hard real
+//! time allocator in the two-level segregated fit family.
+//!
+//! Allocations are divided into two categories: *regular* and *large*. Regular
+//! allocations go into one of the shared slabs managed by an allocator. Large
+//! allocations get their own individual slabs. Due to platform limitations, on
+//! WebGL 2 all vertex buffers are considered large allocations that get their
+//! own slabs.
+//!
+//! The purpose of packing meshes together is to reduce the number of times
+//! vertex and index buffers have to be re-bound, which is expensive.
+//!
+//! [`offset-allocator`]: https://github.com/pcwalton/offset-allocator/
+//! [Sebastian Aaltonen's `OffsetAllocator`]: https://github.com/sebbbi/OffsetAllocator
+
+use std::{
+    fmt::{self, Debug, Display, Formatter},
+    hash::{DefaultHasher, Hash, Hasher},
+    iter,
+    sync::{Arc, RwLock},
+};
+
+use bevy_app::{App, Plugin};
+use bevy_derive::{Deref, DerefMut};
+use bevy_ecs::{
+    schedule::IntoSystemConfigs,
+    system::{ResMut, Resource},
+};
+use bevy_time::common_conditions::on_timer;
+use bevy_utils::{hashbrown::HashMap, prelude::default, tracing::error, Duration};
+use offset_allocator::{Allocation, Allocator};
+use slotmap::{new_key_type, SlotMap};
+use wgpu::{
+    util::BufferInitDescriptor, BufferDescriptor, BufferUsages, DownlevelFlags, IndexFormat,
+};
+
+use crate::{
+    mesh::MeshVertexBufferLayoutRef,
+    render_resource::Buffer,
+    renderer::{RenderAdapter, RenderDevice, RenderQueue},
+    Render, RenderApp,
+};
+
+/// How often we sweep unused allocations.
+const SWEEP_INTERVAL: Duration = Duration::from_secs(10);
+
+/// The default size of a slab, in bytes.
+const DEFAULT_SLAB_SIZE: u64 = 32 * 1024 * 1024;
```

Comment:
Should this (in the future) be synced up with wgpu's allocation sizes as well? (https://github.com/gfx-rs/wgpu/pull/5875)